ibtk/src/utilities/fortran/averaging.f.m4

```diff
@@ -0,0 +1,115 @@
+c
+c     Compute the arithmetic average based on the twelve inputs
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+      REAL function a_avg12(a0,a1,a2,a3,
+     &                      a4,a5,a6,a7,
+     &                      a8,a9,a10,a11)
+      implicit none
+      REAL a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11
+      a_avg12 = (1.d0/12.d0)*(a0+a1+a2+a3+a4+a5+a6+a7+a8+a9+a10+a11)
+      return
+      end
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+c     Compute the harmonic average based on the two inputs
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+      REAL function h_avg2(a0,a1)
+      implicit none
+      REAL a0,a1
+      REAL dmr,nmr, eps
+      nmr = 2.d0
+      eps = epsilon(0.d0)
+c     If any of the arguments is zero, then harmonic average is zero
+      if (abs(a0) .le. sqrt(eps) .or. abs(a1) .le. sqrt(eps)) then
+        h_avg2 = 0.d0
+      else
+        dmr = 1.d0/a0+1.d0/a1
+        h_avg2 = nmr/dmr
+      endif
+      return
+      end
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+c     Compute the harmonic average based on the four inputs
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+      REAL function h_avg4(a0,a1,a2,a3)
+      implicit none
+      REAL a0,a1,a2,a3
+      REAL dmr,nmr,eps
+      nmr = 4.d0
+      eps = epsilon(0.d0)
+
+c     If any of the arguments is zero, then harmonic average is zero
+      if (abs(a0) .le. sqrt(eps) .or. abs(a1) .le. sqrt(eps) .or.
+     &    abs(a2) .le. sqrt(eps) .or. abs(a3) .le. sqrt(eps)) then
+        h_avg4 = 0.0
+      else
+        dmr = 1.d0/a0+1.d0/a1+1.d0/a2+1.d0/a3
+        h_avg4 = nmr/dmr
+      endif
+      return
+      end
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+c     Compute the harmonic average based on the twelve inputs
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+      REAL function h_avg12(a0,a1,a2,a3,
+     &                      a4,a5,a6,a7,
+     &                      a8,a9,a10,a11)
+      implicit none
+      REAL a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11
+      REAL dmr,nmr,eps
+      nmr = 12.d0
+      eps = epsilon(0.d0)
+
+c     If any of the arguments is zero, then harmonic average is zero
+      if (abs(a0) .le. sqrt(eps) .or. abs(a1) .le. sqrt(eps) .or.
+     &    abs(a2) .le. sqrt(eps) .or. abs(a3) .le. sqrt(eps) .or.
+     &    abs(a4) .le. sqrt(eps) .or. abs(a5) .le. sqrt(eps) .or.
+     &    abs(a6) .le. sqrt(eps) .or. abs(a7) .le. sqrt(eps) .or.
+     &    abs(a8) .le. sqrt(eps) .or. abs(a9) .le. sqrt(eps) .or.
+     &    abs(a10) .le. sqrt(eps) .or. abs(a11) .le. sqrt(eps)) then
+          h_avg12 = 0.d0
+      else
+          dmr = 1.d0/a0+1.d0/a1+1.d0/a2+1.d0/a3+1.d0/a4+1.d0/a5
+     &             +1.d0/a6+1.d0/a7+1.d0/a8+1.d0/a9+1.d0/a10+1.d0/a11
+          h_avg12 = nmr/dmr
```

Comment:
@boyceg I don't think its feasible to add something like this as a macro.

If we are worried about performance then we might want to investigate turning on link-time optimization. That would be a more permanent (and performant) fix.